src/Lean/Server/FileWorker/RequestHandling.lean

```diff
@@ -414,6 +414,87 @@ def handleSemanticTokensRange (p : SemanticTokensRangeParams)
   let endPos := text.lspPosToUtf8Pos p.range.end
   handleSemanticTokens beginPos endPos
 
+partial def handleFoldingRange (p : FoldingRangeParams)
+  : RequestM (RequestTask (Array FoldingRange)) := do
+  let doc ← readDoc
+  asTask do
+    let ⟨snaps, e?⟩ ← doc.allSnaps.updateFinishedPrefix
+    let mut stxs := snaps.finishedPrefix.map (·.stx)
+    match e? with
+    | some ElabTaskError.aborted =>
+      throw RequestError.fileChanged
+    | some (ElabTaskError.ioError e) =>
+      throw (e : RequestError)
+    | _ => pure ()
+
+    let lastSnap := snaps.finishedPrefix.getLastD doc.headerSnap
+    stxs := stxs ++ (← parseAhead doc.meta.text.source lastSnap).toList
```

Comment:
This is copied from `handleDocumentSymbol`, right? Let's put that in a new function... hmm... `getDocumentSyntaxWithParseAhead`?

Reply:
Interesting, inaccurate in what way?

> as we'd need to handle editors which don't support partial results

Which is all of them, unfortunately

Reply:
Here is a relatively minimal case where parse-ahead fails. I can only reproduce when running under a debug build, though I imagine the same would happen with a release build if you had a term which takes sufficiently long to elaborate.

```lean
def foo :=
  0

infixl:65 " :=: " => Eq

theorem bar : 0 :=: 0 :=
  rfl
```

Editing the definition of `foo` will result in only `foo` having a folding region as, I assume, `infixl` hasn't been processed and so the definition of `bar` doesn't parse. I don't think there's much that can really be done about this :(.